crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,174 @@
+    let mapped_source = source.mapped_source();
+    let start = mapped_source.map_index(location.offset());
+    let end = mapped_source.map_index(location.end_offset());
+    Some((location.file_id, start..end))
+}
+
+/// Builder for [`DiagnosticReport`]s.
+pub struct ReportBuilder {
+    sources: SourceMap,
+    colors: ColorGenerator,
+    report: ariadne::ReportBuilder<'static, MappedSpan>,
+}
+
+impl ReportBuilder {
+    fn new(sources: SourceMap, location: Option<NodeLocation>) -> Self {
+        let (file_id, range) = location
+            .and_then(|location| map_span(&sources, location))
+            .unwrap_or((FileId::NONE, 0..0));
+        Self {
+            sources,
+            colors: ColorGenerator::new(),
+            report: ariadne::Report::build(ReportKind::Error, file_id, range.start),
+        }
+    }
+
+    /// Set the main message for the report.
+    pub fn with_message(&mut self, message: impl ToString) {
+        self.report.set_message(message);
+    }
+
+    /// Set the help message for the report, usually a suggestion on how to fix the error.
+    pub fn with_help(&mut self, help: impl ToString) {
+        self.report.set_help(help);
+    }
+
+    /// Set a note for the report, providing additional information that isn't related to a
+    /// source location (when a label should be used).
+    pub fn with_note(&mut self, note: impl ToString) {
+        self.report.set_note(note);
+    }
+
+    /// Add a label at a given location. If the location is `None`, the message is discarded.
+    pub fn with_label_opt(&mut self, location: Option<NodeLocation>, message: impl ToString) {
+        if let Some(mapped_span) = location.and_then(|location| map_span(&self.sources, location)) {
+            self.report.add_label(
+                ariadne::Label::new(mapped_span)
+                    .with_message(message)
+                    .with_color(self.colors.next()),
+            );
+        }
+    }
+
+    /// Enable or disable colors in the output.
+    pub fn with_color(self, color: bool) -> Self {
+        let Self {
+            sources,
+            colors,
+            report,
+        } = self;
+        let report = report.with_config(ariadne::Config::default().with_color(color));
+        Self {
+            sources,
+            colors,
+            report,
+        }
+    }
+
+    /// Return the report.
+    pub fn finish(self) -> DiagnosticReport {
+        DiagnosticReport {
+            sources: self.sources,
+            report: self.report.finish(),
+        }
+    }
+}
+
+/// A diagnostic report that can be printed to a CLI with pretty colours and labeled lines of
+/// GraphQL source code.
+pub struct DiagnosticReport {
+    sources: SourceMap,
+    report: ariadne::Report<'static, MappedSpan>,
+}
+
+impl DiagnosticReport {
+    /// Returns a builder for creating diagnostic reports.
+    ///
+    /// Provide GraphQL source files and the main location for the diagnostic.
```

Comment:
Maybe mention that `sources` is typically from `document_or_schema.sources.clone()`, and `location` from `most_relevant_name_or_node.location()`?